platform/security/idp/security-idp-plugin/security-idp-plugin-namequalifier/src/test/groovy/org/codice/ddf/security/idp/plugin/namequalifier/NameQualifierPresignPluginSpec.groovy

```diff
@@ -0,0 +1,182 @@
+
+    def NEGATIVE_TEST = 'negative test'
+
+    def setupSpec() {
+        OpenSAMLUtil.initSamlEngine()
+    }
+
+    def setup() {
+
+        // set up authNReq
+        authNReq.getIssuer() >> issuer
+
+        // set up issuer
+        response.getIssuer() >> issuer
+
+        // set up subject
+        assertion.getIssuer() >> issuer
+
+        // set up basic response
+        response.assertions >> assertions
+        //assertions.add(assertion)
+
+        // set up the name id policy
+        authNReq.getNameIDPolicy() >> nameIdPolicy;
+
+    }
+
+     def 'test name qualifier when issuer format is not null and is persistent'() {
+
+        setup:
+        response.getIssuer().getFormat() >> format
+        assertions.add(assertion)
+        assertion.getSubject() >> subject
+        assertion.getSubject().getNameID() >> nameID
+        nameID.getFormat() >> PERSISTENT
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        1 * nameID.getFormat()
+
+    }
+
+    def 'test name qualifier when no assertions'() {
+
+        setup:
+        response.getIssuer().getFormat() >> format
+        assertion.getSubject() >> subject
+        assertion.getSubject().getNameID() >> nameID
+        nameID.getFormat() >> PERSISTENT
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        0 * nameID.setNameQualifier(NAME_QUALIFIER)
+
+    }
+
+
+    def 'test name qualifier when issuer format is null'() {
+
+        setup:
+        response.getIssuer().getFormat() >> null
+        assertions.add(assertion)
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        0 * nameID.getFormat()
+
+    }
+
+    def 'test name qualifier when issuer format is not persistent'() {
+
+        setup:
+        response.getIssuer().getFormat() >>  NEGATIVE_TEST
+        assertions.add(assertion)
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        0 * nameID.getFormat()
+
+    }
+
+    def 'test name qualifier when assertion subject is null'() {
+
+        setup:
+        response.getIssuer().getFormat() >> format
+        assertions.add(assertion)
+        assertion.getSubject() >> null
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        0 * nameID.getFormat()
+
+    }
+
+    def 'test name qualifier when assertion subject is not null and name id is persistent'() {
+
+        setup:
+        response.getIssuer().getFormat() >> format
+        assertions.add(assertion)
+        assertion.getSubject() >> subject
+        assertion.getSubject().getNameID() >> nameID
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        1 * nameID.getFormat()
```

Comment:
❓ Shouldn't this run a second time for the second half of the condition? 
❓ Should this also affirm that `1 * nameID.setNameQualifier(NAME_QUALIFIER)`?